### PR TITLE
Redefines `Data.Nat.Base._≤″_`

### PR DESCRIPTION
**Pull request by jamesmckinna**
As per issue #1919 (cf. #1726 )

One puzzle is that I (ended up thinking that I had to) introduce a definition equivalent to the former field name projection `proof`, only to then deprecate it immediately. Couldn't quite see how else to trigger a warning about the redefinition of the relation (otherwise, with the single proof modification in `Data.Nat.Properties`, the redefinition actually affects nothing in the library, thanks to the additional pattern synonym!)

Also fixes a spurious whitespace violation from PR #1926 (not sure how this happened).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -427,6 +427,32 @@ Non-backwards compatible changes
...
+    field
+      {k}   : ℕ
+      proof : m + k ≡ n
+  ```
+  which introduced a spurious additional definition, when this is in fact, modulo
+  field names and implicit/explicit qualifiers, equivalent to the definition of left-
+  divisibility, `_∣ˡ_` for the `RawMagma` structure of `_+_`. Since the addition of
+  raw bundles to `Data.X.Base`, this definition can now be made directly. Knock-on
+  consequences include the need to retain the old constructor name, now introduced
+  as a pattern synonym, and deprecation of (a function equiavlent to) the former
```
Comment (JacquesCarette) [inline CHANGELOG.md:445]: typo: equivalent

File: src/Data/Nat/Base.agda [hunk 2]
```diff
@@ -314,51 +362,17 @@ compare (suc m) (suc n) with compare m n
...
-  { Carrier = _
-  ; _≈_ = _≡_
-  ; _+_ = _+_
-  ; _*_ = _*_
-  ; 0# = 0
-  ; 1# = 1
-  }
+-- Version 2.0
 
+proof : ∀ {m n} (le : m ≤″ n) → let less-than-or-equal {k} _ = le in m + k ≡ n
```
Comment (JacquesCarette) [inline src/Data/Nat/Base.agda:374]: So I really don't like exposing this. It wasn't there before, as it really was `_≤″_.proof` which doesn't "pollute" the name space nearly as much. I would seriously consider not defining it at all?

File: src/Data/Nat/Properties.agda [hunk 3]
```diff
@@ -2098,8 +2098,8 @@ _>″?_ = flip _<″?_
 
 ≤‴⇒≤″ : ∀{m n} → m ≤‴ n → m ≤″ n
 ≤‴⇒≤″ {m = m} ≤‴-refl     = less-than-or-equal {k = 0} (+-identityʳ m)
-≤‴⇒≤″ {m = m} (≤‴-step x) = less-than-or-equal (trans (+-suc m _) (_≤″_.proof ind)) where
-  ind = ≤‴⇒≤″ x
+≤‴⇒≤″ {m = m} (≤‴-step x) with less-than-or-equal proof ← ≤‴⇒≤″ x
+  = less-than-or-equal (trans (+-suc m _) proof)
```
Comment (JacquesCarette) [inline src/Data/Nat/Properties.agda:2102]: can't you use proj_2 instead of `with` here?

**Review by JacquesCarette — APPROVED** [review]
It's a tough call, whether to do everything "by hand" using the least possible assumptions, or to reveal a bit more of the algebraic structure that "make things work".

In a sense, I think what I'd want to do, in some hypothetical Agda 5.0, would be to use meta-programming to have the library definition by as general as possible, but the user-facing one be as concrete as possible.

In the meantime, I think this experiment is worthwhile.

File: src/Data/Nat/Properties.agda (hunk 3 above)
Comment (jamesmckinna) [reply src/Data/Nat/Properties.agda:2102]: Ah!... nice question: if you do use `proj₂`, then you are exposing the (new) representation as a Σ-type; whereas, by pattern matching on the synonym, you preserve the original abstraction. This also arises (I realised later) in the very definition of the function `proof` that I introduce for the sake of having a name to deprecate...

File: src/Data/Nat/Base.agda (hunk 2 above)
Comment (jamesmckinna) [reply src/Data/Nat/Base.agda:374]: I agree, but I wasn't sure how to handle deprecation of a field name... even though there was only a single use of it elsewhere in the library, and which is now removed in any case in favour of the abstraction-preserving appeal to the `less-than-or-equal` pattern synonym.

**Comment by jamesmckinna** [comment]
Aaaargh! Committed to wrong branch! Will revert last two commits...

**Comment by jamesmckinna** [comment]
So... I think this is ready to go once we have resolved the issue of whether it is worth the trouble of introducing a functional replacement for the former field projection, solely for the sake of the deprecation warning. It *is* a wart, to be sure.

File: CHANGELOG.md (hunk 1 above)
Comment (gallais) [inline CHANGELOG.md:445]:
```suggestion
  as a pattern synonym, and deprecation of (a function equivalent to) the former
```

**Review by gallais — APPROVED** [review]

**Comment by jamesmckinna** [comment]
The only outstanding issue for me is whether to include the function-instead-of-record-field `proof`, in order to then deprecate it. Any opinions on this, @MatthewDaggitt ? Otherwise, i think it should be `status:ready`...

**Comment by jamesmckinna** [comment]
Can we do this for v2.0?

**Comment by jamesmckinna** [comment]
Aaaargh! That's a horrible (and entirely unwittingly introduced) bug, sorry!
Is it this particular PR, or an Agda/`emacs` main issue?

**Pull request merged**